### PR TITLE
Migrate pallet-lottery to pallet attribute macro

### PR DESCRIPTION
**Pull request by shaunxw**
Part of #7882.

Converts the `Lottery` pallet to the new pallet attribute macro introduced in #6877.

Following the upgrade guidelines here: https://crates.parity.io/frame_support/attr.pallet.html#upgrade-guidelines.

## ⚠️ Breaking Change ⚠️ 

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `Lottery` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `Lottery` pallet.

### CODE REVIEW

File: frame/lottery/src/lib.rs [hunk 1]
```diff
@@ -267,17 +326,18 @@ decl_module! {
...
-		fn start_lottery(origin,
+		#[pallet::weight(T::WeightInfo::start_lottery())]
+		pub(crate) fn start_lottery(
+			origin: OriginFor<T>,
 			price: BalanceOf<T>,
 			length: T::BlockNumber,
 			delay: T::BlockNumber,
 			repeat: bool,
-		) {
+		) -> DispatchResultWithPostInfo {
```
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:336]:
```suggestion
		) -> DispatchResult {
```

File: frame/lottery/src/lib.rs [hunk 2]
```diff
@@ -288,74 +348,36 @@ decl_module! {
...
 				Ok(())
 			})?;
 			// Make sure pot exists.
 			let lottery_account = Self::account_id();
 			if T::Currency::total_balance(&lottery_account).is_zero() {
 				T::Currency::deposit_creating(&lottery_account, T::Currency::minimum_balance());
 			}
-			Self::deposit_event(RawEvent::LotteryStarted);
+			Self::deposit_event(Event::<T>::LotteryStarted);
+			Ok(().into())
```
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:360]:
```suggestion
			Ok()
```

File: frame/lottery/src/lib.rs [hunk 3]
```diff
@@ -244,17 +302,18 @@ decl_module! {
 		/// provided by this pallet, which uses storage to determine the valid calls.
 		///
 		/// This extrinsic must be called by the Manager origin.
-		#[weight = T::WeightInfo::set_calls(calls.len() as u32)]
-		fn set_calls(origin, calls: Vec<<T as Config>::Call>) {
+		#[pallet::weight(T::WeightInfo::set_calls(calls.len() as u32))]
+		pub(crate) fn set_calls(origin: OriginFor<T>, calls: Vec<<T as Config>::Call>) -> DispatchResultWithPostInfo {
```
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:306]:
```suggestion
		pub(crate) fn set_calls(origin: OriginFor<T>, calls: Vec<<T as Config>::Call>) -> DispatchResult {
```

File: frame/lottery/src/lib.rs [hunk 4]
```diff
@@ -244,17 +302,18 @@ decl_module! {
...
-				CallIndices::kill();
+				CallIndices::<T>::kill();
 			} else {
 				let indices = Self::calls_to_indices(&calls)?;
-				CallIndices::put(indices);
+				CallIndices::<T>::put(indices);
 			}
-			Self::deposit_event(RawEvent::CallsUpdated);
+			Self::deposit_event(Event::<T>::CallsUpdated);
+			Ok(().into())
```
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:316]:
```suggestion
			Ok(())
```

File: frame/lottery/src/lib.rs [hunk 5]
```diff
@@ -227,15 +284,16 @@ decl_module! {
...
 		///
 		/// This extrinsic must be called by a signed origin.
-		#[weight =
+		#[pallet::weight(
 			T::WeightInfo::buy_ticket()
 				.saturating_add(call.get_dispatch_info().weight)
-		]
-		fn buy_ticket(origin, call: Box<<T as Config>::Call>) {
+		)]
+		pub(crate) fn buy_ticket(origin: OriginFor<T>, call: Box<<T as Config>::Call>) -> DispatchResultWithPostInfo {
```
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:291]:
```suggestion
		pub(crate) fn buy_ticket(origin: OriginFor<T>, call: Box<<T as Config>::Call>) -> DispatchResult {
```

File: frame/lottery/src/lib.rs [hunk 6]
```diff
@@ -227,15 +284,16 @@ decl_module! {
...
 			let caller = ensure_signed(origin.clone())?;
 			call.clone().dispatch(origin).map_err(|e| e.error)?;
 
 			let _ = Self::do_buy_ticket(&caller, &call);
+			Ok(().into())
```
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:296]:
```suggestion
			Ok(())
```

File: frame/lottery/src/lib.rs [hunk 7]
```diff
@@ -288,74 +348,36 @@ decl_module! {
...
 		}
 
 		/// If a lottery is repeating, you can use this to stop the repeat.
 		/// The lottery will continue to run to completion.
 		///
 		/// This extrinsic must be called by the `ManagerOrigin`.
-		#[weight = T::WeightInfo::stop_repeat()]
-		fn stop_repeat(origin) {
+		#[pallet::weight(T::WeightInfo::stop_repeat())]
+		pub(crate) fn stop_repeat(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
```
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:368]:
```suggestion
		pub(crate) fn stop_repeat(origin: OriginFor<T>) -> DispatchResult {
```

File: frame/lottery/src/lib.rs [hunk 8]
```diff
@@ -288,74 +348,36 @@ decl_module! {
...
-							return T::WeightInfo::on_initialize_end()
-						}
-						// We choose not need to kill Participants and Tickets to avoid a large number
-						// of writes at one time. Instead, data persists between lotteries, but is not used
-						// if it is not relevant.
-					}
-				}
-				return T::DbWeight::get().reads(1)
-			})
+			Ok(().into())
```
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:375]:
```suggestion
			Ok(())
```

**Review by shawntabrizi — APPROVED** [review]

File: frame/lottery/src/lib.rs (hunk 8 above)
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:375]:
```suggestion
			Ok(())
```

File: frame/lottery/src/lib.rs (hunk 7 above)
Comment (shawntabrizi) [inline frame/lottery/src/lib.rs:368]:
```suggestion
		pub(crate) fn stop_repeat(origin: OriginFor<T>) -> DispatchResult {
```

**Comment by shaunxw** [comment]
I think `DispatchResultWithPostInfo` is also used in some merged migration PRs from me. Will make a PR and replace them with `DispatchResult`.

**Review by shawntabrizi — APPROVED** [review]

**Review by thiolliere — APPROVED** [review]

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**